reasoning_gym/arc/rearc.py

```diff
@@ -0,0 +1,185 @@
+    """
+    prefix = "verify_"
+    return {strip_prefix(n, prefix): getattr(verifiers, n) for n in dir(verifiers) if n.startswith(prefix)}
+
+
+@dataclass
+class ReArcConfig:
+    diff_lb: int = 0
+    diff_ub: int = 1
+    board_format_opts: BoardFormattingOptions = field(default_factory=default_board_format_opts)
+    seed: Optional[int] = None
+    size: int = 500
+
+    def validate(self):
+        assert self.diff_lb < self.diff_ub, "diff_lb must be < diff_ub."
+        assert self.size > 0, "Size of dataset must be positive."
+
+
+class ReArcDataset(ProceduralDataset):
+    def __init__(self, config: ReArcConfig):
+        super().__init__(config=config, seed=config.seed, size=config.size)
+        self.board_format_opts = config.board_format_opts
+        self._prompt_templates = _REARC_PROMPT_TEMPLATES
+        self.diff_lb = config.diff_lb
+        self.diff_ub = config.diff_ub
+        self._generators = get_generators()
+        self._verifiers = get_verifiers()
+
+    @staticmethod
+    def get_rng_difficulty(rng: Random) -> float:
+        if not hasattr(rng, "difficulty_samples"):
+            return 0.0
+        samples = rng.difficulty_samples
+        avg = sum(samples) / len(samples) if samples else 0.0
+        rng.difficulty_samples = []
+        return avg
+
+    @staticmethod
+    def get_pso_difficulty(example: dict) -> float:
+        """
+        PSO-Difficulty: proxy measure for example difficulty, defined as weighted sum of #Pixels, #Symbols, #Objects
+        """
+        i, o = example["input"], example["output"]
+        hwi = height(i) * width(i)
+        hwo = height(o) * width(o)
+        pix_pct = (hwi + hwo) / 1800
+        col_pct = len(palette(i) | palette(o)) / 10
+        obj_dens = (len(objects(i, T, F, F)) / hwi + len(objects(o, T, F, F)) / hwo) / 2
+        return (pix_pct + col_pct + obj_dens) / 3
+
+    def __len__(self) -> int:
+        return self.size
+
+    @staticmethod
+    def visualise_pair(example: Dict[str, Any]) -> None:
+        """
+        Visualise a ReArc task pair
+        """
+        norm = Normalize(vmin=0, vmax=9)
+        args = {"cmap": _COLOUR_MAP, "norm": norm}
+
+        # Change to 1 row, 2 columns
+        height = 1
+        width = 2
+        figure_size = (3 * width * 3, height * 3)
+        figure, axes = plt.subplots(height, width, figsize=figure_size)
+
+        # Plot input and output side by side
+        axes[0].imshow(example["metadata"]["input"], **args)
+        axes[1].imshow(example["metadata"]["output"], **args)
+
+        # Add titles to distinguish the plots
+        axes[0].set_title("Input")
+        axes[1].set_title("Output")
+
+    def format_rearc_input(self, idx: int, task: dict, generator: Callable) -> str:
+        """
+        Format a ReArc task input with multiple examples and test input.
+        """
+        example_1 = generator(Random((self.seed + idx) * 1 * self.size), self.diff_lb, self.diff_ub)
```

Comment:
Could you make the number of shown examples per task a configuration parameter (e.g. in range min_training_examples, max_training_examples), maybe with defaults 3, 6? if you look at examples in the notebook in https://github.com/michaelhodel/re-arc/blob/main/demo.ipynb I think hard coded 3 examples would be a bit low.